# eBay Car Sales Analysis


The aim of this project is to clean the data and analyze the included used car listings. We will start by importing the libraries we need and reading the dataset into pandas.


In [1]:
import pandas as pd
file = 'autos.csv'
autos = pd.read_csv('/Users/nstanzione/Documents/EDU/DataQuest/Data/' + file)

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observations

The total dataset we are working with has 50,000 rows and 20 columns. The headings all have camelcase instead of our preferred snakecase. There are only 4 fields that are int types (two of which are related to date-time). There are 6 fields with NULL values (tho never a majority of the field values. All other fields are object type. At high-level, here are some of the initial data issues:

* Price has unit indicator (eg. $) and commas
* Odometer has commas and labels (e.g. km)
* Some data may be foreign langugae, or appears to have mispellings (e.g. "manuell")
* abTest is not a boolean type


## Task 1: Update Columns
Let's first start by updating the column names to snakecase.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
cols = ['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen']

cols_clean = cols
cols_clean[7] = 'registration_year'
cols_clean[12] = 'registration_month'
cols_clean[15] = 'unrepaired_damage'
cols_clean[16] = 'ad_created'
cols_clean[17] = 'num_pictures'

print(cols_clean)


['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model', 'odometer', 'registration_month', 'fuelType', 'brand', 'unrepaired_damage', 'ad_created', 'num_pictures', 'postalCode', 'lastSeen']


In the code above we idenitifed those names we wanted to change manually due to the fact they had multiple upper case words. Next, we will look to create two functions that we can use to update the remaining columns.

In [6]:
def first(string):
    for i in range(0,len(string)):
        if (string[i].istitle()):
            return i

    return 0
    
def replace_cap(string):
    n = first(string)
    string_new = string[:n] + '_' + string[n].lower()+string[n+1:].lower()
    return string_new

test = 'testRun'
print(replace_cap(test))
    

test_run


The test above shows that our two new functions will be able to help us update the remaining column names. The first function has a string as an input, and then identifies the character in the string that has the first uppercase letter, and returns its position. The second function has a string as an input, that string is then fed through the first function, and the position of the upper case character is then used to adjust the string accordingly to be snakecase. We will loop through the remaining "bad" field names below and update the columns accordingly.

In [7]:
bad_labels = [0,3,6,9,13,18,19]

for n in bad_labels:
    cols_clean[n] = replace_cap(cols_clean[n])

print(cols_clean)

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code', 'last_seen']


In [8]:
autos.columns = cols_clean
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

As the new dataframe headings show, we were able to successfully update the field names from camelcase to snakecase using the two functions described above after manually updating a few specific fields.

Now, let's explore the dataset to see what else may need to be updated.

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In reviewing the dataframe, it looks like 'seller' and 'offer_type' are all the same value except for one record. This does not provide much value so we should drop these columns.

Other columns needing review:
* price is not stored as a number
* odometer is not stored as number
* vehicle_type has the most common type as limousine
* unrepaired_damage & gearbox appears to be in foreign language

## Task 2: Updating price and odometer values to numbers

In [10]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

For the odometer column, we are only dealing with a few values. We know that we only need to remove the 'km' and ',' from the existing values and then convert the series to the float data type.

In [11]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)


autos['odometer'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer, dtype: int64

An initial look at the price column, it appears we will only need to be concerned with '$' and commas. 

In [12]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(float)

autos['price'].value_counts()

0.0        1421
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
6202.0        1
18310.0       1
898.0         1
11240.0       1
789.0         1
Name: price, Length: 2357, dtype: int64

In [13]:
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)
autos.rename(columns={'price' : 'price_usd'}, inplace=True)

In [14]:
del autos['seller']
del autos['offer_type']

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   price_usd           50000 non-null  float64
 3   abtest              50000 non-null  object 
 4   vehicle_type        44905 non-null  object 
 5   registration_year   50000 non-null  int64  
 6   gearbox             47320 non-null  object 
 7   power_ps            50000 non-null  int64  
 8   model               47242 non-null  object 
 9   odometer_km         50000 non-null  int64  
 10  registration_month  50000 non-null  int64  
 11  fuel_type           45518 non-null  object 
 12  brand               50000 non-null  object 
 13  unrepaired_damage   40171 non-null  object 
 14  ad_created          50000 non-null  object 
 15  num_pictures        50000 non-null  int64  
 16  post

As seen above, the data is much more manageable and price and odometer date are now float values and able to be analyzed further. 

## Task 3: Analyze Price & Odometer
Let's dig into whether there are any unusual data points in these two fields.

In [15]:
autos['odometer_km'].unique().shape

(13,)

In [16]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [17]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The odometer data is fairly boring in that there does not seem to be any issues or outliers. There are 13 unique records, which means we have rounded values. The range is reaosnable from 5K to 150K and each value has multiple records. Let's move onto price.

In [18]:
autos['price_usd'].unique().shape

(2357,)

In [19]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [20]:
autos['price_usd'].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64

In [21]:
autos['price_usd'].value_counts(ascending=False).head()

0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
Name: price_usd, dtype: int64

The price column is a different story. From the information above there looks to be several imperfections in the data:
* 1,421 records with 0 price
* Several records above 350K price with oddities with bad data entry (e.g. 12,345,678) or outliers (e.g. 99,999,999). We will work under the assumption that no person is spending over 350K on a car on ebay and remove these values.

In [22]:
autos = autos[autos['price_usd'].between(1,400000)]
autos['price_usd'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

In [23]:
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

As seen above, the minimum price for a car on eBay was 1.  The average car sold for ~6K and 50% of cars sold between 1200 and 7500.

As for the odemeter readings, 50% of the cars were over 150K km on the odometer. This leads us to know that this might have been the maximum input value for the data entry. The average care had 125K km and the lowest reading was 5K.

### Task 4: Clean-up & Analysis of Date Fields
Right now, the date_crawled, last_seen, and ad_created columns are all identified as the 'object' type by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

In [24]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [25]:
autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['ad_created'] = autos['ad_created'].str[:10]
autos['last_seen'] = autos['last_seen'].str[:10]

autos[['date_crawled','ad_created','last_seen']][0:5]

<ipython-input-25-c0d632329825>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['date_crawled'] = autos['date_crawled'].str[:10]
<ipython-input-25-c0d632329825>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['ad_created'] = autos['ad_created'].str[:10]
<ipython-input-25-c0d632329825>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.

date_crawled  ad_created   last_seen
0   2016-03-26  2016-03-26  2016-04-06
1   2016-04-04  2016-04-04  2016-04-06
2   2016-03-26  2016-03-26  2016-04-06
3   2016-03-12  2016-03-12  2016-03-15
4   2016-04-01  2016-04-01  2016-04-01

In [26]:
autos['date_crawled'].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The date crawled data is very evenly spread, each date accounts for roughly 3% of the data. The crawler looks to have begun in March 5, 2016. There are lower percentage of the data has a date_crawled value for the later dates 4/5/2016 through 4/7/2016. This means there have been fewer posts crated the past three dates.

In [32]:
autos['ad_created'].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The data above shows that a larger percentage of the records have been created in 2016. This makes sense as the older posts find buyers and are replaced with new posts of cars-for-sale. However, it appears that there are no dates with signifcantly higher postings. The distribution above shows each day in the month of March 2016 having roughly 3% percent of the total ad postings. This shows that there is a steady stream of posts each day, and the website might be delayed in recording that date information as the three most recent dates seem to have a lower percentage or they could have a review process before making the actual posting.

In [28]:
autos['last_seen'].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The last_seen date distribution is highly skewed to the later dates. This seems reasonable as most ads will likely be removed after they are sold. The crawler has seen ~50% of the data within the past three days. 

In [33]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There is definitely incorrect data values in the registion_year field particularly the min & max values. Also, the registration_year seems to represent something different than the US meaning. For the US, any registration_year value that is not the current or prior year would be considered incorrect since you have to update your car registartion every year. This may be the original car purchase date.

In [41]:
print(autos['registration_year'].value_counts().sort_index().head(7))
print(autos['registration_year'].value_counts().sort_index().tail(15))

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
Name: registration_year, dtype: int64
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64


The first registration_year that seems reasonable is 1910. The final year is a bit more tricky, there are 1,400 records with a 2017 registration year. Many car manufactors advertise the current year vehicle release as the next year's model. There are even 470 records with a 2018 date, two years in the future! Let's take a look at a few examples of these 2017 and 2018 cars for sale.

In [47]:
autos[autos['registration_year']==2017].head(10)

date_crawled                                               name  \
84     2016-03-27                                     Renault_twingo   
164    2016-03-13  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
390    2016-03-25   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
802    2016-03-19                 Lada_mit_wenig_km_neuem_Tuev_bj_08   
918    2016-03-15                              BMW_114i_Sondermodell   
1121   2016-03-26  Opel_Zafira_A_1.8_Z18XE_HU_02/18_164Tkm_Alu_Klima   
1551   2016-03-08            Verkaufe_hier_meine_Auto_vw_Passat_2007   
1889   2016-03-11                                        Golf_2_90PS   
1903   2016-04-05                               Tausche_Ford_Granada   
2006   2016-03-15                        Renault_Twingo_1.2_TÜV_neue   

      price_usd   abtest vehicle_type  registration_year    gearbox  power_ps  \
84        900.0  control          NaN               2018        NaN        60   
164      4800.0  control          NaN               2018    manuell         0   
390      7750.0     test          NaN               2018    manuell        76   
802      2100.0     test          NaN               2018    manuell         0   
918     14900.0  control          NaN               2018    manuell       102   
1121     1650.0  control          NaN               2018    manuell       125   
1551     3200.0     test          NaN               2018    manuell       140   
1889     1950.0  control          NaN               2018  automatik        90   
1903     1111.0     test          NaN               2018        NaN         0   
2006     1799.0  control          NaN               2018    manuell        57   

       model  odometer_km  registration_month fuel_type       brand  \
84    twingo       150000                   0       NaN     renault   
164   meriva        80000                   4    benzin        opel   
390   andere       150000                   6    benzin        fiat   
802   kalina       150000                   0    benzin        lada   
918      1er        40000                   3    benzin         bmw   
1121  zafira       150000                   9       NaN        opel   
1551  passat       150000                   0    diesel  volkswagen   
1889    golf        70000                  10    benzin  volkswagen   
1903  andere        70000                   0    benzin        ford   
2006  twingo         5000                   7       NaN     renault   

     unrepaired_damage  ad_created  num_pictures  postal_code   last_seen  
84                 NaN  2016-03-27             0        40589  2016-04-05  
164               nein  2016-03-13             0        37627  2016-04-04  
390               nein  2016-03-25             0        78239  2016-03-28  
802                NaN  2016-03-19             0        12621  2016-03-21  
918               nein  2016-03-15             0        24850  2016-03-22  
1121               NaN  2016-03-26             0        95466  2016-04-06  
1551               NaN  2016-03-08             0        29392  2016-03-15  
1889              nein  2016-03-11             0        89155  2016-03-24  
1903               NaN  2016-04-05             0        23623  2016-04-07  
2006               NaN  2016-03-15             0        64297  2016-03-16

In [51]:
autos[autos['registration_year']==2018].head(10)

date_crawled                                               name  \
84     2016-03-27                                     Renault_twingo   
164    2016-03-13  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
390    2016-03-25   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
802    2016-03-19                 Lada_mit_wenig_km_neuem_Tuev_bj_08   
918    2016-03-15                              BMW_114i_Sondermodell   
1121   2016-03-26  Opel_Zafira_A_1.8_Z18XE_HU_02/18_164Tkm_Alu_Klima   
1551   2016-03-08            Verkaufe_hier_meine_Auto_vw_Passat_2007   
1889   2016-03-11                                        Golf_2_90PS   
1903   2016-04-05                               Tausche_Ford_Granada   
2006   2016-03-15                        Renault_Twingo_1.2_TÜV_neue   

      price_usd   abtest vehicle_type  registration_year    gearbox  power_ps  \
84        900.0  control          NaN               2018        NaN        60   
164      4800.0  control          NaN               2018    manuell         0   
390      7750.0     test          NaN               2018    manuell        76   
802      2100.0     test          NaN               2018    manuell         0   
918     14900.0  control          NaN               2018    manuell       102   
1121     1650.0  control          NaN               2018    manuell       125   
1551     3200.0     test          NaN               2018    manuell       140   
1889     1950.0  control          NaN               2018  automatik        90   
1903     1111.0     test          NaN               2018        NaN         0   
2006     1799.0  control          NaN               2018    manuell        57   

       model  odometer_km  registration_month fuel_type       brand  \
84    twingo       150000                   0       NaN     renault   
164   meriva        80000                   4    benzin        opel   
390   andere       150000                   6    benzin        fiat   
802   kalina       150000                   0    benzin        lada   
918      1er        40000                   3    benzin         bmw   
1121  zafira       150000                   9       NaN        opel   
1551  passat       150000                   0    diesel  volkswagen   
1889    golf        70000                  10    benzin  volkswagen   
1903  andere        70000                   0    benzin        ford   
2006  twingo         5000                   7       NaN     renault   

     unrepaired_damage  ad_created  num_pictures  postal_code   last_seen  
84                 NaN  2016-03-27             0        40589  2016-04-05  
164               nein  2016-03-13             0        37627  2016-04-04  
390               nein  2016-03-25             0        78239  2016-03-28  
802                NaN  2016-03-19             0        12621  2016-03-21  
918               nein  2016-03-15             0        24850  2016-03-22  
1121               NaN  2016-03-26             0        95466  2016-04-06  
1551               NaN  2016-03-08             0        29392  2016-03-15  
1889              nein  2016-03-11             0        89155  2016-03-24  
1903               NaN  2016-04-05             0        23623  2016-04-07  
2006               NaN  2016-03-15             0        64297  2016-03-16

Seeing as the odometer reading for these cars is well over 50K, we can be certain these are not "newer" vehicles. However, this two years represent 4% of our entire dataset, so we will not remove the records entirely. However, let's review the distirbution of those records with "correct" registration years. 

In [60]:
autos_reg = autos[autos['registration_year'].between(1910,2016)]
autos_reg['registration_year'].value_counts(normalize=True).sort_index().tail(25)

1992    0.007926
1993    0.009104
1994    0.013474
1995    0.026285
1996    0.029412
1997    0.041794
1998    0.050620
1999    0.062060
2000    0.067608
2001    0.056468
2002    0.053255
2003    0.057818
2004    0.057904
2005    0.062895
2006    0.057197
2007    0.048778
2008    0.047450
2009    0.044665
2010    0.034040
2011    0.034768
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, dtype: float64

Majority of the car registrations are from 1997 to 2010. This seems reasonable since eBay should be mostly leveraged for used car sales. A very small percentage of cars for sale are pre-1990 (16+ years old).

### Task 5: Brand Review
Let's dig a bit further into the brands most commonly sold on eBay based on our dataset.

In [71]:
autos['brand'].value_counts()

volkswagen        10336
opel               5277
bmw                5274
mercedes_benz      4652
audi               4168
ford               3382
renault            2325
peugeot            1430
fiat               1262
seat                920
skoda               780
nissan              741
mazda               739
smart               694
citroen             686
toyota              611
hyundai             483
sonstige_autos      471
volvo               439
mini                418
mitsubishi          399
honda               388
kia                 345
alfa_romeo          321
porsche             287
suzuki              286
chevrolet           275
chrysler            169
dacia               129
daihatsu            122
jeep                108
subaru              103
land_rover           99
saab                 79
daewoo               76
jaguar               74
trabant              68
rover                65
lancia               55
lada                 29
Name: brand, dtype: int64

The most popular brands are all European, further enforcing that the dat we are reviewing is foreign-based. Volkswagen is the top brand followed by Opel, and then the 3 popular luxury brands (BMW, Mercedes, Audi). Lada is the least common brand. 

Looking at the above list of brands, we will explore the top-10 brands as they account for ~80% of our data.

In [74]:
autos['brand'].value_counts().index[:10]

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')

In [81]:
top10_brands = autos['brand'].value_counts().index[:10]
brand_price_avg = {}

for b in top10_brands:
    temp_df = autos[autos['brand']==b]
    mean_price = temp_df['price_usd'].mean()
    brand_price_avg[b] = mean_price

print(brand_price_avg)


{'volkswagen': 5332.4784249226, 'opel': 2944.6075421641085, 'bmw': 8261.382442169132, 'mercedes_benz': 8536.027085124677, 'audi': 9212.9306621881, 'ford': 3728.4121821407452, 'renault': 2431.195698924731, 'peugeot': 3065.611888111888, 'fiat': 2793.8700475435817, 'seat': 4315.744565217391}


As expected, our three luxury brands BMW, Mercedes, and Audi have the highest average sale price. All three have an average between 8K-10K USD (with Audi being thr most expensive. The Renault has the lowest average price at ~2,400 USD. Our most common brand, Volkswagen, is in the middle of the pack at ~5,300 USD.

In [82]:
brand_odometer_avg = {}

for b in top10_brands:
    temp_df = autos[autos['brand']==b]
    mean_odometer = temp_df['odometer_km'].mean()
    brand_odometer_avg[b] = mean_odometer

print(brand_odometer_avg)

{'volkswagen': 128896.57507739938, 'opel': 129383.17225696419, 'bmw': 132682.97307546454, 'mercedes_benz': 130796.43164230438, 'audi': 129492.56238003839, 'ford': 124349.49733885274, 'renault': 128062.36559139784, 'peugeot': 127356.64335664336, 'fiat': 117567.35340729002, 'seat': 121641.30434782608}


In [87]:
avg_price = pd.Series(brand_price_avg)
avg_odometer = pd.Series(brand_odometer_avg)

top_brands = pd.DataFrame(avg_price,columns=['avg_price_usd'])
top_brands['avg_odometer_km'] = avg_odometer

pd.options.display.float_format = '{:,.0f}'.format

top_brands

avg_price_usd  avg_odometer_km
volkswagen             5,332          128,897
opel                   2,945          129,383
bmw                    8,261          132,683
mercedes_benz          8,536          130,796
audi                   9,213          129,493
ford                   3,728          124,349
renault                2,431          128,062
peugeot                3,066          127,357
fiat                   2,794          117,567
seat                   4,316          121,641

There is no clear distinction of the mileage acorss the top-10 brands. If we rememebr from our earlier analysis, the odometer data might be a bit misleading with a potential maximum input value of 150,000. There is not logical correlation present between average mileage and average price for the brands. The brand itself is much more determinate of the price.